### ISSUE TITLE
Splatting-based Synthesis for Video Frame Interpolation

### ISSUE THREAD

**Issue by DavidePaglieri** [sniklaus/softmax-splatting#51]
Hello, well done on your new paper, I found it really interesting! Do you plan on releasing the source code of the corrected softmax-splatting operator and the new pipeline?

**Comment by sniklaus** [comment]
Thanks for the kind words! I do want to release the stable softmax splatting operator, yes. I just need to make sure that I follow my employer's rules to do so since they own the intellectual property of it. This may take a while though, so stay tuned. :+1:

**Comment by DavidePaglieri** [comment]
Thank you, looking forward to it! Also, where can I find the supplementary material of the paper? It is mentioned on arXiv but I cannot find it.

**Comment by sniklaus** [comment]
You can find an interactive demo here: http://sniklaus.com/splatsyn
A side-by-side comparison with SoftSplat: http://content.sniklaus.com/splatsyn/sidebyside.mp4
Some representative 4K results: http://content.sniklaus.com/splatsyn/4k.mp4

**Comment by DavidePaglieri** [comment]
Can I also ask you what alpha values did you use for the merge map and splat map while training end-to-end? I am trying to find them myself but currently having some trouble.

**Comment by sniklaus** [comment]
We set these six alpha values to 1 initially and then learn their actual values through end-to-end training. I just pulled the resulting alphas from our trained model and you can find the summary below.

- splat_photo: 2.35
- splat_flow: 5.89
- splat_varia: 1.24
- merge_photo: 164.96
- merge_flow: 108.04
- merge_varia: 1.18

**Comment by DavidePaglieri** [comment]
Thank you for your reply! There are still a few things that are not 100% clear to me.

I am trying to work out what I am doing wrong in reproducing your results. Using your values I don't get sensible merge and splat maps in the way I calculate them. I suspect this in particular is due to the flow variance, which is close to zero almost everywhere, so doing 1/(1+merge_varia*varia) gives values close to 1 which makes the overall map become almost completely white.

This is the way I calculate them

```
def splat_image_and_map(self, image_0, image_1, flow_0_1, flow_1_0, t):

    photo = torch.linalg.norm(image_0 - self.backwarp(tenInput=image_1, tenFlow=flow_0_1), dim=1, ord=2)
    flow = torch.linalg.norm(flow_0_1 + self.backwarp(tenInput=flow_1_0, tenFlow=flow_0_1), dim=1, ord=2)
    varia = torch.sqrt(torch.linalg.norm( (self.G(torch.square(flow_0_1))-torch.square(self.G(flow_0_1))) , dim=1, ord=2))
    
    merge_map = (1.0/(1.0+self.merge_p*photo) + 1.0/(1.0+self.merge_f*flow) + 1.0/(1.0+self.merge_v*varia)).unsqueeze(1)
    splat_map = (1.0/(1.0+self.splat_p*photo) + 1.0/(1.0+self.splat_f*flow) + 1.0/(1.0+self.splat_v*varia)).unsqueeze(1)
    
    flow_0_t = flow_0_1 * t
    
    t_centric_flow = softsplat.softsplat(tenIn=-flow_0_t, tenFlow=flow_0_t, tenMetric=splat_map, strMode='soft')
    
    image_t = self.backwarp(tenInput=image_0, tenFlow=t_centric_flow)
    merge_map_t = softsplat.softsplat(tenIn=merge_map, tenFlow=flow_0_t, tenMetric=splat_map, strMode='soft').clip(0.001, 1.0)
    
    return image_t, merge_map_t
```

I have moved the sqrt of the flow variance outside the norm as it was giving some NaN values due to taking the square root of very small negative numbers (otherwise I could also clip them to 0). Could you tell me if I am calculating these values correctly?

If I remove the contribution of the flow variance to the merge/splat maps, and manually set the remaining alpha values to high numbers I get something like this, which is somewhat closer to the merge map you had in the paper.
![dap-merge](https://user-images.githubusercontent.com/30180090/154294263-66041b07-68b6-47ac-a5ef-94a94e354afd.png)

Another problem I am having is that I cannot seem to make the values converge properly, as the PSNR on vimeo90k stays close to 31 dB without ever increasing significantly. I am using: 

- batch size: 32
- AdaMax optimiser
- learning rate: 0.0001
- betas=(0.9, 0.999)

If I increase the learning rate too much the weights of the PWC-Net diverge and the optical flow becomes meaningless. But with such a low learning rate it takes an eternity to converge. Could you please share more details on how you are training it?
Thank you for your help!

**Comment by sniklaus** [comment]
Apologies in advance for not replying in great detail, I am a little swamped at the moment.

Try normalizing the images before computing the metrics such that they have zero mean and unit standard deviation. Try using the L_1 instead of the L_2 norm for computing the metrics. Try using .clip(0.0, None) before applying the sqrt. Try gradient clipping, training PWC-Net is subject to noisy gradients in my experience.

And be aware that the parameters I stated in my previous message were initialized to 1 and end-to-end trained. They hence probably work well with the flow estimator that I am using but maybe not as well with others. Also, you are not using Gaussian splatting but it should still converge okay without it.

**Comment by DavidePaglieri** [comment]
Thank you! Do you mean normalising the flows with zero mean and unit std as well? I am using PWC-Net as well for flow estimation at the moment.

**Comment by sniklaus** [comment]
I left the flows as is, only the images. If you normalize the flows for the warping then the warping is wrong, and I tried normalizing the flows before computing the flow variance but that didn't improve the results.

**Comment by DavidePaglieri** [comment]
Yeah, I was referring to normalizing the flows for the flow consistency and variance maps, but it didn't make any difference when I tried it either.

Last question I promise. The following is a video of 8x slow-mo on the car_turn, I am using GMA instead of PWC-Net in this case, I haven't yet done the complete end-to-end training on vimeo90k, and I am not yet using the gaussian splatting for the t-centric optical flow. Except for that, I am using your pipeline.

https://user-images.githubusercontent.com/30180090/154810317-d450851e-19a8-465f-9110-907b44dc6fa9.mp4

There are three main problems that I can see here:

1. The pixelated artefacts at the front and back of the car, which I suppose are due to the fact that I don't have the gaussian splatting implementation working yet.
2. The vertical lines at the far left and right of the frames, which are either due to the same problem as above or they probably get fixed with end-to-end training
3. The black holes between the car windows. The backwarped images at the same frames don't have these holes, instead, they are created when combining the two images with the forward splatted merge maps. Are these artefacts also due to the fact that I am not using the Gaussian kernel? Are they supposed to go away after the end-to-end training of the flow estimator with the merge/splat maps? Do you have these artefacts in your videos as well? 

Overall though, it's amazing how fast this frame interpolation method is compared to other state-of-the-art algorithms! I really like it!

**Comment by sniklaus** [comment]
I am afraid the video doesn't play for me, I tried both my computer and my phone. Could you try sharing it again? Or screenshots of the issues that you are describing? Thanks!

**Comment by DavidePaglieri** [comment]
Sorry about that, try with this one

https://drive.google.com/file/d/1JZrVHCZxfpTsk2UZ1PJl9tb6E_rxRPLe/view?usp=sharing

**Comment by sniklaus** [comment]
That video works, thanks!

The vertical bars are because the flow is sampling a pixel at the boundary of the image and due to bilinear sampling it will average valid pixels inside the image with invalid pixels from outside of it. You can somewhat alleviate that by changing the padding_mode of grid_sample which is zero by default.

I am afraid that I don't have a good idea where the other artifacts originate from, I don't think it is the Gaussian splatting though. Are you currently training end to end, meaning that you also learn the six alpha parameters?

**Comment by DavidePaglieri** [comment]
Sorry for the late reply. With end-to-end training it did work a bit better, but some of the artifacts weren't completely gone. Are you still planning on releasing the stable gaussian splatting operator? 

I have one more question regarding the original softmax splatting paper. In the videos did you perform recursive interpolation at 0.5 or did you do time modulated interpolation? I have found that the same model performs more smoothly when interpolating recursively instead of when interpolating frames at an arbitrary time. Although, the latter is faster because you don't have to recompute the optical flow everytime and only need the warping and the fusion net instead

**Comment by sniklaus** [comment]
I still want to release it, but since my implementation is property of Adobe it is not necessarily straightforward to do so. So my apologies that I do not have an estimate of when this will happen (it may take a long time).

I did not perform recursive interpolation and all results from methods that support an arbitrary interpolation instant where utilized as such. Otherwise the performance benefits for multi-frame interpolation would not apply anymore since the optical flow would have to be estimated over and over again (as you pointed out as well as well).

**Comment by DavidePaglieri** [comment]
Okay I see. By the way, I believe that the white-ish banding artifacts that show up in the video that I sent here, are the same artifacts that your new paper with Ping Hu solves with many-to-many splatting, which is definitely another very interesting approach. I would be very curious to see how those two appraches compare on 8x interpolation on a few Full HD and 4K videos, and also how they compare to FILM...

**Comment by fransiskusyoga** [comment]
Hello Simon Niklaus. I also tried to replicate splatting-based paper but I still get a different result. I get computed average psnr 32.49940574435201, computed average ssim 0.9640149. This is still far from 33.5 PSNR that you got. Here some detail of my setup
- I use your PWCNet https://github.com/sniklaus/pytorch-pwc . I run your PWCNet "estimate" function in pwcnet.run
- I test it on Vimeo90K from this GitHub page https://github.com/feinanshan/M2M_VFI/blob/main/Test/bench_vimeo90k.py
- here is how I calculate the phi
``` python
def calculate_phi(flow_a, flow_b, img_a, img_b):
    # brightness consistenc
    phi_photo = img_a - warp(img_b, flow_a)
    phi_photo = torch.sum(phi_photo*phi_photo, axis=1).unsqueeze(1)
    phi_photo = torch.sqrt(phi_photo)

    # flow consistency
    phi_flow = flow_a + warp(flow_b, flow_a)
    phi_flow = torch.sum(phi_flow*phi_flow, axis=1).unsqueeze(1)
    phi_flow = torch.sqrt(phi_flow)

    # variance value
    flow_a_square_conv = flow_a*flow_a
    kernel = torch.tensor([
        [1.0/16, 1.0/8, 1.0/16],
        [1.0/8 , 1.0/4, 1.0/8 ],
        [1.0/16, 1.0/8, 1.0/16]]).view(1,1,3,3).repeat([2,1,1,1]).to(flow_a.device)
    flow_a_square_conv = torch.nn.functional.conv2d(flow_a_square_conv, kernel, padding=1, groups=2)
    flow_a_conv_square = torch.nn.functional.conv2d(flow_a, kernel, padding=1, groups=2)
    flow_a_conv_square = flow_a_conv_square*flow_a_conv_square
    phi_varia = flow_a_square_conv - flow_a_conv_square
    phi_varia = torch.sqrt(phi_varia*phi_varia)
    phi_varia = torch.sum(phi_varia, axis=1).unsqueeze(1)

    return phi_photo, phi_flow, phi_varia
```
I used your alpha value  (splat_photo: 2.35, splat_flow: 5.89, splat_varia: 1.24, merge_photo: 164.96, merge_flow: 108.04, merge_varia: 1.18). In the snippet above I use L2 distance but I also have tried L1 distance and it does not improve a lot.
- and this is how I do forward warp
```python
    def forward(self, img0, img1, flow_01, flow_10, t):
        # calculate weight for splatting and merging
        phi_photo_01, phi_flow_01, phi_varia_01 = calculate_phi(flow_01, flow_10, img0, img1)
        w_01_splat = 1 / (1 + self.alpha_splat[0] * phi_photo_01)+ \
                     1 / (1 + self.alpha_splat[1] * phi_flow_01 )+ \
                     1 / (1 + self.alpha_splat[2] * phi_varia_01) 
        w_01_merge = 1 / (1 + self.alpha_merge[0] * phi_photo_01)+ \
                     1 / (1 + self.alpha_merge[1] * phi_flow_01 )+ \
                     1 / (1 + self.alpha_merge[2] * phi_varia_01) 
        phi_photo_10, phi_flow_10, phi_varia_10 = calculate_phi(flow_10, flow_01, img1, img0)
        w_10_splat = 1 / (1 + self.alpha_splat[0] * phi_photo_10)+ \
                     1 / (1 + self.alpha_splat[1] * phi_flow_10 )+ \
                     1 / (1 + self.alpha_splat[2] * phi_varia_10) 
        w_10_merge = 1 / (1 + self.alpha_merge[0] * phi_photo_10)+ \
                     1 / (1 + self.alpha_merge[1] * phi_flow_10 )+ \
                     1 / (1 + self.alpha_merge[2] * phi_varia_10) 
        
        flow_t0 = gaussian_forwardwarp(-t*flow_01, t*flow_01, w_01_splat)
        flow_t1 = gaussian_forwardwarp(-(1-t)*flow_10, (1-t)*flow_10, w_10_splat)
        w_01_merge = gaussian_forwardwarp(w_01_merge, t*flow_01, w_01_splat)
        w_10_merge = gaussian_forwardwarp(w_10_merge, (1-t)*flow_10, w_10_splat)
        img_t0 = warp(img0, flow_t0)
        img_t1 = warp(img1, flow_t1)

        # final merging
        tw_01 = t*w_01_merge
        tw_10 = (1-t)*w_10_merge
        total = torch.clip(tw_01 + tw_10, min = 0.0000001) # prevent divide by 0
        tw_01 = tw_01/total
        tw_10 = tw_10/total
        img = (img_t0*tw_01 + img_t1*tw_10)
        return img
```
here is some more detail 1). warp() is just a standard backward warp, 2). gaussian_forwardwarp() is softmax forward splatting but after the splatting, I applied gaussian kernel convolution. I hope gaussian_forwardwarp() will be the same as your gaussian forward warp. (let me know if it is not the same). Hand here is the result. I tried different kernel sizes (3,5,7,11) with a gaussian standard deviation of 1 pixel. 
- What are the steps that I did wrong?
